Doc/library/devmode.rst

```diff
@@ -0,0 +1,213 @@
+.. _devmode:
+
+Python Development Mode
+=======================
+
+.. versionadded:: 3.7
+
+The Python Development Mode introduces additional runtime checks that are too
+expensive to be enabled by default. It should not be more verbose than the
+default if the code is correct: new warnings are only emitted when an issue is
+detected.
+
+It can be enabled using the :option:`-X dev <-X>` command line option or by
+setting the :envvar:`PYTHONDEVMODE` environment variable to ``1``.
+
+Effects of the Python Development Mode
+======================================
+
+Enabling the Python Development Mode is similar to the following command, but
+with additional effects described below::
+
+    PYTHONMALLOC=debug PYTHONASYNCIODEBUG=1 python3 -W default -X faulthandler
+
+Effects of the Python Development Mode:
+
+* Add ``default`` :ref:`warning filter <describing-warning-filters>`. The
+  following warnings are shown:
+
+  * :exc:`DeprecationWarning`
+  * :exc:`ImportWarning`
+  * :exc:`PendingDeprecationWarning`
+  * :exc:`ResourceWarning`
+
+  Normally, the following warnings are filtered by the default :ref:`warning
+  filters <describing-warning-filters>`.
+
+  It behaves as if the :option:`-W default <-W>` command line option is used.
+
+  Use the :option:`-W error <-W>` command line option or set
+  the :envvar:`PYTHONWARNINGS` environment variable to ``error`` to treat warnings
+  as errors.
+
+* Install debug hooks on memory allocators to check for:
+
+  * Buffer underflow
+  * Buffer overflow
+  * Memory allocator API violation
+  * Usage of the GIL
```

Comment:
Regarding "Usage of the GIL", should that specifically be "Unsafe usage of the GIL"? For example, do the debugs hooks trigger when:

1) Correctly using the `PyInterpreterState`, `PyThreadState`, and `PyGilState` APIs, exactly as the [C-API documentation describes](https://docs.python.org/3/c-api/init.html#thread-state-and-the-global-interpreter-lock)?

2) Unsafely/incorrectly using the GIL?

If it's (1), the current version is correct, but if it's (2), I'd recommend changing it to the following:

```suggestion
  * Unsafe usage of the GIL
```

The current version seems to imply that attempting to access or use the GIL in *any* form triggers the debug hooks. Is this true?